sdk-extensions/tracing-incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/trace/ExecutorServiceSpanProcessor.java

```diff
@@ -0,0 +1,210 @@
+            scheduleDelayNanos,
+            maxExportBatchSize,
+            exporterTimeoutNanos,
+            new ArrayBlockingQueue<>(maxQueueSize),
+            SPAN_PROCESSOR_TYPE_LABEL,
+            SPAN_PROCESSOR_TYPE_VALUE,
+            executorService,
+            isShutdown,
+            workerScheduleIntervalNanos);
+    this.ownsExecutorService = ownsExecutorService;
+    this.executorService = executorService;
+    executorService.schedule(worker, workerScheduleIntervalNanos, TimeUnit.NANOSECONDS);
+  }
+
+  @Override
+  public void onStart(Context parentContext, ReadWriteSpan span) {}
+
+  @Override
+  public boolean isStartRequired() {
+    return false;
+  }
+
+  @Override
+  public void onEnd(ReadableSpan span) {
+    if (!span.getSpanContext().isSampled()) {
+      return;
+    }
+    worker.addSpan(span);
+  }
+
+  @Override
+  public boolean isEndRequired() {
+    return true;
+  }
+
+  @Override
+  public CompletableResultCode shutdown() {
+    if (isShutdown.getAndSet(true)) {
+      return CompletableResultCode.ofSuccess();
+    }
+    CompletableResultCode result = worker.shutdown();
+    // do the cleanup after worker finishes flush
+    result.whenComplete(
+        () -> {
+          if (ownsExecutorService) {
+            executorService.shutdown();
+          }
+        });
+
+    return result;
+  }
+
+  @Override
+  public CompletableResultCode forceFlush() {
+    return worker.forceFlush();
+  }
+
+  // Visible for testing
+  List<SpanData> getBatch() {
+    return new ArrayList<>(worker.getBatch());
+  }
+
+  private static class Worker extends WorkerBase {
+
+    private final AtomicLong nextExportTime = new AtomicLong();
+
+    private final ArrayBlockingQueue<SpanData> batch;
+
+    private final AtomicBoolean isShutdown;
+
+    private final long workerScheduleIntervalNanos;
+
+    private Worker(
+        SpanExporter spanExporter,
+        long scheduleDelayNanos,
+        int maxExportBatchSize,
+        long exporterTimeoutNanos,
+        BlockingQueue<ReadableSpan> queue,
+        String spanProcessorTypeLabel,
+        String spanProcessorTypeValue,
+        ScheduledExecutorService executorService,
+        AtomicBoolean isShutdown,
+        long workerScheduleIntervalNanos) {
+      super(
+          spanExporter,
+          scheduleDelayNanos,
+          maxExportBatchSize,
+          exporterTimeoutNanos,
+          queue,
+          spanProcessorTypeLabel,
+          spanProcessorTypeValue,
+          executorService);
+
+      this.isShutdown = isShutdown;
+      this.batch = new ArrayBlockingQueue<>(maxExportBatchSize);
+      this.workerScheduleIntervalNanos = workerScheduleIntervalNanos;
+      updateNextExportTime();
+    }
+
+    private void updateNextExportTime() {
+      nextExportTime.set(System.nanoTime() + scheduleDelayNanos);
+    }
+
+    @Override
+    public void run() {
+      // nextExportTime is set for the first time in the constructor
+
+      continueWork.set(true);
```

Comment:
@Oberon00 I don't know what would happen if the next run of the worker `Runnable` will be scheduled on a different thread than the previous one. If it is ok to have a local boolean variable, I will update the code.

Reply:
But it looks like the boolean is never used outside this method, except in shutdown, and there it seems to change nothing as you already check isShutdown too. But I may have misread the logic.

Reply:
`isShutdown` - should the SpanProcessor shutdown. If true, then no more work should be done.
`continueWork` - should the current run (loop) continue or not. If false, then at some point in the future the loop will be running again.

Reply:
The question is: If `continueWork.set(false);` is done in WorkerBase.shutdown, is `isShutdown` also already true? If so, then that access to continueWork could be removed. And then we have the situation that at the beginning of `run` continueWork is always set before it is read and continueWork is never used outside the method. In that case it can be transformed without change of behavior to a local boolean.

Reply:
@Oberon00 you are correct. I've converted `continueWork` to local variable.